### ISSUE TITLE
Light up pad of Novation Launchpad S

### ISSUE THREAD

**Issue by blablack** [monocasual/giada#36]
It would be great if I could set up Giada to midi output the status of devices like the Novation Launchpad S to show the status of a channel, especially wait/waiting, off and playing (in the novation launchpad, it would correspond to blinking, off, on).

**Comment by blablack** [comment]
I created a fork of Giada on my github account - https://github.com/blablack/giada

I own a Novation Launchpad S - I would be happy enough to do the dev.
Novation Launchpad S can receive simple midi messages, so it should "simply" be a matter of sending the right messages when needed. With a proper design this could be generic enough to be compatible with other similar devices.

**Comment by monocasual** [comment]
Very good, code submission and hacking are always welcome here :) What kind of midi messages wants the Launchpad S? Sysex or something more generic?

**Comment by blablack** [comment]
The Launchpad S wants simple midi messages - Note On + Velocity 127 to light up, Note On + Velocity 0 to turn off, etc.

I have a small "proof of concept" that works for the Launchpad S but I want to polish it first before commiting it here: https://github.com/blablack/giada/tree/midi_lightning
Once I have it working for the Launchpad S, I want to make the code generic for other devices (like the MPC for example).

I have actually would have one question about this:
I can see two solutions on how to
support mapping for different devices and what exact messages needs to
be sent to light up, blink, etc.
1. The Bitwig way - each devices is hardcoded. The pros is that it is
very easy for the users: they select their devices if it is supported
otherwise it's simply not supported. The cons is that if their devices
is not supported, they would have to ask a dev and/or add the support
in the Giada's code themselves.
2. The Luppp way - Giada support config files for the mapping. The
pros is that Giada can then support any devices as long as someone
created the map file. The cons is that creating the map is quite
complicated for the common user.

I personally prefer option number 1 - it makes the code easier and
from experience messing around with config files although it gives
more freedom....well it's a mess.

Do you have any preferences and/or feedback?

**Comment by ViktorNova** [comment]
What about a combination of 1 and 2? Giada could read config mappings from a system directory AND a user directory, this way the config files that come with Giada work out of the box, while at the same time users can easily define their own. It would be cool to include some little blurb in Giada pointing users to a website or github page for more info on how to make their own config..just an idea

This is very cool!

**Comment by monocasual** [comment]
The combination of 1 and 2 sounds very promising. We provide the "most popular" device settings and let the users write their own setup in the public config page. @blablack : what if we put eveything in a configuration file, without hardcoding the parameters in the code?

**Comment by blablack** [comment]
All right, so the plan is:

1. Have a hard-coded proof of concept to have Giada to send status for Playing, Mute and Solo to the Launchpad S.
This is now done here https://github.com/blablack/giada/tree/midi_lightning
@monocasual if you have the chance to have a look at the code that would be great

2. Create a config file for the Launchpad S and change the hard-coded part to read the messages from the config file.

3. Modify the Midi Configuration GUI to read the available config files and be able to select the midi device used (at the moment Generic and Launchpad S).

4. Finally be able for users to load their own config file.


If feedback there is, please let me know :)

**Comment by monocasual** [comment]
Very good @blablack , thanks for your time! Feel free to issue another pull request with your improvements :+1:

**Comment by blablack** [comment]
Step 2. of the plan is now implemented in my branch:
https://github.com/blablack/giada/tree/midi_lightning

If you get the chance to check, especially the format of the config file that would be great!

I won't do a pull request until I have the 4 points implemented.

**Comment by monocasual** [comment]
@blablack I just browsed throught your branch: if I understand correctly, you're basically sending MIDI messages to the Launchpad with `kernelMidi::midi_*` methods. That's the right way to do it (at least in Giada's realm :) ).

I like the neatness of the .giadamap config file, I guess the string (e.g.) `0:0x90nn3F00` contains channels and messages joined into a single value. Just a question: could you provide a template of the string format? Something like [channel]:[midi_?]nn[midi_?].

A side note: if you wish you can extend the config reader from dataStorage, a class made for reading configuration files made of key=value parameters. 
Keep up the good work!

**Comment by blablack** [comment]
@monocasual : Yes it's exactly that:
- first part is the channel
- second part is the full midi messages, where Giada will replace 'nn' by the note.

In the case of the Launchpad S, they "hijack" the midi messages to decide what the message means, for example, velocity represents the color, last bit is for double buffering, etc.
So this format should allow for other devices if they put the note number in another position.

Hope I'm clear enough!

**Comment by monocasual** [comment]
@blablack I've just pulled your code to take a look around, everything seems clear! Unfortunately at the moment I cannot test it with a real "blinking" device, I'll check it with a MIDI monitor. :+1:

**Comment by blablack** [comment]
I'm still working on providing a bundle of midi map with Giada and give the possibility for the user to load his/her own map.

By the way, where should the bundle be installed by Giada ? I guess /usr/share/giada for Linux, but what about Windows/Mac?

**Comment by monocasual** [comment]
@blablack good question. What if we put the bundle near giada.conf (i.e. `~/.giada/midimaps` or something similar) for Linux? The same also for Mac: `[home]/Library/Application Support/Giada/midimaps`. On Windows we could place the folder alongside the executable. What do you say?

**Comment by blablack** [comment]
So, I have a pull request for this here: https://github.com/monocasual/giada/pull/44

It's not 100% complete, I still need to fix the default path where the bundled midi map are stored.

**Comment by blablack** [comment]
Hi all, 

To me the functionality is finished but needs testing!
I can only test it with my Novation Launchpad in Linux.

I added more explanations on the pull request how things are supposed to work.

If you need anything else from me, please let me know!

**Comment by monocasual** [comment]
Hi @blablack , 

We are in the middle of a massive merge of your branch into master. I have a question: now if I click on a MIDI channel -> "setup MIDI output" I get the following window:

![](http://i.imgur.com/r3Lddfi.png)

Which is different from the original one:

![](http://i.imgur.com/cXkKtVo.png)

Is it ok? If so, what is it for?

**Comment by blablack** [comment]
Hi,

Yes it is OK.

The new window enables the user to chose which button will be light up for
the 3 actions.

For example if you click 'learn' and click a button on the Launchpad, that
button will now react/light up to the playing status.

I use it a lot to have only one button be red for example when any samples
on a channel is muted.

I hope I'm clear :-) I would love to be able to create a quick video of
this but never got the chance.

**Comment by monocasual** [comment]
> The new window enables the user to chose which button will be light up for
> the 3 actions.

Ok great! :+1:  Basically you bind an event (play/ending, mute, solo) to a specific button. I still have a couple of questions: 

* the drop down menu for the output channel is now gone. Is it right?
* shouldn't we enable that kind of midi-lightning binding for sample channels as well?

**Comment by blablack** [comment]
> the drop down menu for the output channel is now gone. Is it right?
Yes, the channel should be caught during the learning phase

> should we enable that kind of midi-lightning binding for sample channels as well?
It's coded for sample channels, do you mean midi channels? Yes it should be enabled as well.

**Comment by monocasual** [comment]
> Yes, the channel should be caught during the learning phase

Ok, so I think that the dropdown should stay. I mean, in the master branch (with no midi-lightning merge), the MIDI output setup is used to send MIDI events from a MIDI channel to the outside world. For example, I record some actions in the action editor and I want to output them to another MIDI device. Thus the dropdown menu defines the MIDI output channel (and the checkbox enables the feature).

Now, with the MIDI lightning feature, a MIDI channel has to send two kinds of messages: "internal" events as described above and MIDI lightning messages. The latter cannot take over the former, so I guess the two windows in the pictures above have to be merged. In the same window one box will define the MIDI output, another one will define the MIDI lightning output (plus learning tools).  

In summary:

* MIDI output for *sample channels*: a new window with MIDI lightning tools (checkbox + learning tools)
* MIDI output for *MIDI channels*: redesigned window with old MIDI output (checkbox + drop-down menu) and new MIDI lightnint tools (checkbox + learning tools).

Does it make sense?

**Comment by blablack** [comment]
Yes it does make sense - I don't think I can add anything to this
description :)

**Comment by monocasual** [comment]
Hi @blablack , I guess I have to bother you with another (simple, I hope) question. 

Currently a "lightning message" is structured as `[channel]:[custom-message]`: isn't the `[channel]` part redundant information? Why don't you set the channel number in the 1st byte of the message, e.g. `0x9<chan>...`?

**Comment by monocasual** [comment]
I answer my own dumb question:  you don't know what the 1st byte really means, so you have to store the channel information somewhere else.

**Comment by monocasual** [comment]
Ok I guess the merge is finally complete. We tweaked a couple of things from the original PoC:

1) The makefile process has been removed. Midimap files will be stored in `[your_home]/.giada/midimaps` folder (except for Windows of course). We will start with a minimal set of midimaps hosted on GitHub, while users can upload their own versions here and/or on the official forum;

2) Midimap values are converted into uint32_t variables instead of the previous 4 bytes array, along with the byte offset for the note. Then the final message is merged right before the kernelMidi::send call.

Everything is in the master branch now and will be released soon. Aggressive testing is welcome :D. Thanks again!
